README.md
<p align="center">
  <big><strong>Bcc GraphQL</strong></big>
</p>

<p align="center">
  <img width="200" src=".github/images/bcc-logo.png"/>
</p>

[![CI][img_src_CI]][workflow_CI]

<hr/>

## Overview

Cross-platform, _typed_, and **queryable** API for Bcc. The project contains multiple [packages] for composing 
GraphQL services to meet specific application demands, and a [docker-compose stack] serving the included 
[bcc-graphql-server Dockerfile], the extended [hasura Dockerfile], [bcc-node-ogmios]. The [schema] is defined in
native `.graphql`, and used to generate a [TypeScript package for client-side static typing]. A mutation is available to 
submit a signed and serialized transaction to the local node.
 
[Apollo Server] exposes the NodeJS execution engine over a HTTP endpoint, and includes support for open source metrics
via Prometheus, and implementing operation filtering to deny unexpected queries. Should you wish to have more control
over the server, or stitch the schema with an existing service, consider importing the executable schema from the 
`@bcc-graphql/api-*` packages only.

**GraphQL** is a query language and execution environment with server and client implementations across many programming
languages. The language can be serialized for network transmission, schema implementations hashed for assurance, and is
suited for describing most domains.
 
**TypeScript** (and JS) has the largest pool of production-ready libraries, developers, and interoperability in the
GraphQL and web ecosystem in general. TypeScript definitions for the schema, generated by [GraphQL Code Generator], are
[available on npm].

**[Ogmios]** is a protocol translation service written in Haskell running on top of bcc-node. It offers a JSON
interface through WebSockets and enables applications to speak Shardagnostic' client mini-protocols via remote procedure
calls.

## Getting Started
Check the [releases] for the latest version.
``` console
git clone \
  --single-branch \
  --branch <VERSION> \
  --recurse-submodules \
  https://github.com/The-Blockchain-Company/bcc-graphql.git \
  && cd bcc-graphql
```

### Up
Choose **one** of the following:

#### A) Build and Run via Docker Compose
Boot the [docker-compose stack] using a convention for container and volume scoping based on the network, as well as
optionally hitting the remote cache to speed up the build. The containers are detached, so you can terminate the log
console session freely. See [Docker Compose docs] to tailor for your use-case
 
<details open>
  <summary><i>mainnet</i></summary>

``` console
DOCKER_BUILDKIT=1 \
COMPOSE_DOCKER_CLI_BUILD=1 \
docker-compose up -d --build &&\
docker-compose logs -f
```
</details>

<details>
  <summary><i>testnet</i></summary>

``` console
DOCKER_BUILDKIT=1 \
COMPOSE_DOCKER_CLI_BUILD=1 \
API_PORT=3101 \
HASURA_PORT=8091 \
OGMIOS_PORT=1338 \
POSTGRES_PORT=5433 \
METADATA_SERVER_URI="https://metadata.bcc-testnet.tbcodev.io" \
docker-compose -p testnet up -d --build &&\
docker-compose -p testnet logs -f
```

</details>

<details>
  <summary><i>aurum -purple</i></summary>

``` console
DOCKER_BUILDKIT=1 \
COMPOSE_DOCKER_CLI_BUILD=1 \
API_PORT=3102 \
HASURA_PORT=8092 \
OGMIOS_PORT=1339 \
POSTGRES_PORT=5434 \
METADATA_SERVER_URI="https://metadata.bcc-testnet.tbcodev.io" \
docker-compose -p aurum -purple up -d --build &&\
docker-compose -p aurum -purple logs -f
```

</details>


#### B) Pull and Run via Docker Compose
Pull images from Docker Hub and run using a convention for container and volume scoping based on the network. The
containers are detached, so you can terminate the log console session freely. See [Docker Compose docs] to tailor for
your use-case.

<details open>
  <summary><i>mainnet</i></summary>

``` console
export NETWORK=mainnet &&\
docker pull tbco/bcc-graphql:5.1.0-beta.1-${NETWORK} &&\
docker pull tbco/bcc-graphql-hasura:5.1.0-beta.1 &&\
docker pull tbco/bcc-node-ogmios:v4.0.0-beta.6-${NETWORK} &&\
docker-compose up -d &&\
docker-compose logs -f
```
</details>

<details>
  <summary><i>testnet</i></summary>

``` console
export NETWORK=testnet &&\
docker pull tbco/bcc-graphql:5.1.0-beta.1-${NETWORK} &&\
docker pull tbco/bcc-graphql-hasura:5.1.0-beta.1 &&\
docker pull tbco/bcc-node-ogmios:v4.0.0-beta.6-${NETWORK} &&\
API_PORT=3101 \
HASURA_PORT=8091 \
OGMIOS_PORT=1338 \
POSTGRES_PORT=5433 \
docker-compose -p ${NETWORK} up -d &&\
docker-compose -p ${NETWORK} logs -f
```

</details>

<details>
  <summary><i>aurum -purple</i></summary>

``` console
export NETWORK=aurum -purple &&\
docker pull tbco/bcc-graphql:5.1.0-beta.1-${NETWORK} &&\
docker pull tbco/bcc-graphql-hasura:5.1.0-beta.1 &&\
docker pull tbco/bcc-node-ogmios:v4.0.0-beta.6-${NETWORK} &&\
API_PORT=3102 \
HASURA_PORT=8092 \
OGMIOS_PORT=1339 \
POSTGRES_PORT=5434 \
docker-compose -p ${NETWORK} up -d &&\
docker-compose -p ${NETWORK} logs -f
```

</details>

### Down
The following commands will not remove volumes, however should you wish to do so, append `-v`

<details open>
  <summary><i>mainnet</i></summary>

``` console
docker-compose down
```
</details>

<details>
  <summary><i>testnet</i></summary>

``` console
docker-compose -p testnet down
```

</details>

<details>
  <summary><i>aurum -purple</i></summary>

``` console
docker-compose -p aurum -purple down
```

</details>

### Check Bcc DB sync progress
Use the GraphQL Playground in the browser at http://localhost:3100/graphql:
``` graphql 
{ bccDbMeta { initialized syncPercentage }}
```
or via command line:
``` console
curl \
  -X POST \
  -H "Content-Type: application/json" \
  -d '{"query": "{ bccDbMeta { initialized syncPercentage }}"}' \
  http://localhost:3100/graphql
```
:information_source: _Wait for `initialized` to be `true` to ensure the epoch dataset is complete. After the first sync
you may need to restart the services using `docker-compose restart bcc-graphql` if the GraphQL server isn't
running._

### Query the full dataset
```graphql
{ bcc { tip { number slotNo epoch { number } } } }
```
``` console
curl \
  -X POST \
  -H "Content-Type: application/json" \
  -d '{"query": "{ bcc { tip { number slotNo epoch { number } } } }"}' http://localhost:3100/graphql
```
### :tada:
``` json
{ "data": { "bcc": { "tip": { "number": 4391749, "slotNo": 4393973, "epoch": { "number": 203 } } } } }
```

For more information, have a look at the [Wiki :book:].

## How to install (Linux / Docker)

### Docker

See [Using Docker].

### From Source 

See [Building].

## Documentation

| Link                                                                                               | Audience                                                     |
| ---                                                                                                | ---                                                          |
| [API Documentation]                                                                                | Users of the Bcc GraphQL API                             |
| [Wiki :book:]                                                                                      | Anyone interested in the project and our development process |
| [Example Queries - Bcc DB Hasura]        | Users of the Bcc DB Hasura API                             |

<hr/>

<p align="center">
  <a href="https://github.com/The-Blockchain-Company/bcc-graphql/blob/master/LICENSE"><img src="https://img.shields.io/github/license/The-Blockchain-Company/bcc-graphql.svg?style=for-the-badge" /></a>
</p>

[img_src_CI]: https://github.com/The-Blockchain-Company/bcc-graphql/workflows/CI/badge.svg
[workflow_CI]: https://github.com/The-Blockchain-Company/bcc-graphql/actions?query=workflow%3ACI
[packages]: ./packages
[docker-compose stack]: ./docker-compose.yml
[Docker Compose docs]: https://docs.docker.com/compose/
[bcc-graphql-server Dockerfile]: ./Dockerfile
[hasura Dockerfile]: ./packages/api-bcc-db-hasura/hasura/Dockerfile
[bcc-node-ogmios]: https://ogmios.dev/getting-started/docker/
[schema]: ./packages/api-bcc-db-hasura/schema.graphql
[TypeScript package for client-side static typing]: ./packages/client-ts/README.md
[Apollo Server]: https://www.apollographql.com/docs/apollo-server/
[GraphQL Code Generator]: https://graphql-code-generator.com
[available on npm]: https://www.npmjs.com/package/bcc-graphql-ts
[Ogmios]: https://ogmios.dev/
[releases]: https://github.com/The-Blockchain-Company/bcc-graphql/releases
[Wiki :book:]: https://github.com/The-Blockchain-Company/bcc-graphql/wiki
[Using Docker]: https://github.com/The-Blockchain-Company/bcc-graphql/wiki/Docker
[Building]: https://github.com/The-Blockchain-Company/bcc-graphql/wiki/Building
[API Documentation]: https://The-Blockchain-Company.github.io/bcc-graphql
[Example Queries - Bcc DB Hasura]: ./packages/api-bcc-db-hasura/src/example_queries

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [5.1.0-beta.1](https://github.com/The-Blockchain-Company/bcc-graphql/compare/5.1.0-beta.0...5.1.0-beta.1) (2021-08-15)

### Compatible with:

- [`bcc-node`: `aurum-purple-1.0.1`](https://github.com/The-Blockchain-Company/bcc-node/releases/tag/aurum-purple-1.0.1)
- [`bcc-ogmios`: `v4.0.0-beta.6`](https://github.com/BccSolutions/ogmios/releases/tag/v4.0.0-beta.6)
- [`bcc-db-sync`: `aurum-purple-1.0.1`](https://github.com/The-Blockchain-Company/bcc-db-sync/releases/tag/aurum-purple-1.0.1) - Note: The database must be recreated using the new version.
- [`hasura/graphql-engine`: `1.3.3`](https://github.com/hasura/graphql-engine/releases/tag/v1.3.3)

### Features
* support Docker pull and run to skip builds ([ee03a57](https://github.com/The-Blockchain-Company/bcc-graphql/commit/ee03a57b7045bf3947faa0df1eaa536b3c7c482a))

### Bug Fixes

* invalid cache target in compose file ([4b4dfc3](https://github.com/The-Blockchain-Company/bcc-graphql/commit/4b4dfc37a1a7c4b3729e251219ca0fb25464d8f9))


## [5.1.0-beta.0](https://github.com/The-Blockchain-Company/bcc-graphql/compare/5.0.0...5.1.0-beta.0) (2021-08-12)

### Compatible with:

- [`bcc-node`: `aurum-purple-1.0.1`](https://github.com/The-Blockchain-Company/bcc-node/releases/tag/aurum-purple-1.0.1)
- [`bcc-db-sync`: `aurum-purple-1.0.1`](https://github.com/The-Blockchain-Company/bcc-db-sync/releases/tag/aurum-purple-1.0.1) - Note: The database must be recreated using the new version.
- [`hasura/graphql-engine`: `1.3.3`](https://github.com/hasura/graphql-engine/releases/tag/v1.3.3)

### Features
* feature: Aurum data mapping ([eff1d9c](https://github.com/The-Blockchain-Company/bcc-graphql/commit/eff1d9ced4cdcd2adee6602de801feecc28451bf))
* feature: add `AurumGenesis`([d600dc0](https://github.com/The-Blockchain-Company/bcc-graphql/commit/d600dc01849292fe0a5006202a9be936e56d105b))
* feature: add Aurum blocks to conditional chain following ([8d6aaf4](https://github.com/The-Blockchain-Company/bcc-graphql/commit/8d6aaf46f01eef3774e2f2efa07838c125cfd1fb))

### Bug Fixes
* Guard logic ([110deaa](https://github.com/The-Blockchain-Company/bcc-graphql/commit/110deaaba6ef0bd8b3ecbe28d4979ee26335302c))
* OperationRequiresNodeInCurrentEra -> OperationRequiresSyncedNode ([6165965](https://github.com/The-Blockchain-Company/bcc-graphql/commit/616596568e709731387c65fea7dc1cfa4cacf92d))
* startup race-condition with chain-followers ([0b15315](https://github.com/The-Blockchain-Company/bcc-graphql/commit/0b15315f4084f6cf2de1773e55a9217149c33e4b))


## [5.0.0](https://github.com/The-Blockchain-Company/bcc-graphql/compare/4.0.0...5.0.0) (2021-07-14)

### Compatible with:

- [`bcc-node`: `1.27.0`](https://github.com/The-Blockchain-Company/bcc-node/releases/tag/1.27.0)
- [`bcc-db-sync`: `10.0.1`](https://github.com/The-Blockchain-Company/bcc-db-sync/releases/tag/10.0.1) - Note: The database must be recreated using the new version.
- [`hasura/graphql-engine`: `1.3.3`](https://github.com/hasura/graphql-engine/releases/tag/v1.3.3)

### ⚠ BREAKING CHANGES

* `Asset.assetName` and `Asset.assetId` are now typed as `Hex`, as the relationship from `Assets` to `TokenMints` needed
  to be established using the underlying table, where `assetId` is not present.
* Configuration of the asset metadata fetching is now a single value. `POLLING_INTERVAL_METADATA_SYNC_INITIAL` and
  `POLLING_INTERVAL_METADATA_SYNC_ONGOING` are replaced by `ASSET_METADATA_UPDATE_INTERVAL`, which is the number of
  seconds before the service checks the registry for updates.

* AssetSupply.total is now an optional field, and can return null.
* Transaction.metadata is now `JSON` type, not `JSONObject`
  - Despite the name, `JSONObject` was mapped to the underlying JSON resolver as a workaround to avoid breaking changes.
* The fields previously modelled on Token have been nested under Token.asset
* `tokens` and `tokens_aggregate` have been removed in favour of assets and assets_aggregate
- Asset properties nested under `PaymentAddress.assets` are
  now under `PaymentAddress.assets.asset`
* `Block.merkleRoot` removed, as no longer part of `bcc-db-sync` schema. ([8b3b718](https://github.com/The-Blockchain-Company/bcc-graphql/commit/8b3b7185f619cc6ede85ac18b43c96fdb76ba19e))

### Features
* Replaces the use of `bcc-cli` with [Ogmios](https://github.com/BccSolutions/ogmios/) for interacting with the
  node using JSON-WSP, and implements the convenient `bcc-node-ogmios` Docker image as a drop-in replacement for
  the `bcc-node` image.
* Sync assets using chain-sync protocol via Ogmios ([784509e](https://github.com/The-Blockchain-Company/bcc-graphql/commit/784509eed509dc7bfd856aa4ee9cea7a7b40b046))
* `tokenMints` and `tokenMints_aggregate` queries ([f803b94](https://github.com/The-Blockchain-Company/bcc-graphql/commit/f803b94f67fd016c47d0832510f7af9628a2186c))
* add query complexity validations ([24a14a9](https://github.com/The-Blockchain-Company/bcc-graphql/commit/24a14a9cde019d4232fc9abf5cb4c6dcbf37638f)), 
* add default query complexity calculations with arguments ([b16c77d](https://github.com/The-Blockchain-Company/bcc-graphql/commit/b16c77d0411696736ed3028f53925c8fa320780f))
* implement complexity in resolvers ([5664a55](https://github.com/The-Blockchain-Company/bcc-graphql/commit/5664a5505c52168de0b343ab81f5350a31e99bb3))
* use bcc-node config for lastConfiguredMajorVersion ([5a621c2](https://github.com/The-Blockchain-Company/bcc-graphql/commit/5a621c2d1f16ec90693ee57b1c67d7d56b6ece33)), closes [#454](https://github.com/The-Blockchain-Company/bcc-graphql/issues/454)
* add Epoch_bool_exp.startedAt ([4cc580a](https://github.com/The-Blockchain-Company/bcc-graphql/commit/4cc580accc7eb7ae09e608ab1023e6a70cda1206))

test/postgres/init/001_restore.sh
pg_restore -d bccexplorer -U postgres /docker-entrypoint-initdb.d/bccexplorer.pg_dump

test/postgres/dump_when_block_height_reached.sh
#!/bin/bash

HTTP_SERVER_URI=$1
TARGET_BLOCK_HEIGHT="70000"
DB_NAME=$2
FILE_NAME='bccexplorer.pg_dump'
OUT_PATH="test/postgres/init/${FILE_NAME}"

printf "\nPolling $HTTP_SERVER_URI until block height exceeds $TARGET_BLOCK_HEIGHT\n"
while true;
do
	blockheight=$(curl -s -X POST -H "Content-Type: application/json" -d '{"query": "{ bcc { tip: { number }}"}' "$HTTP_SERVER_URI" | jq ".data.bcc.tip.number");
	if [ ! $blockheight = 'null' ]; then
	  printf "\r$(date +%H:%M:%S): $blockheight";
    if [ "${blockheight}" -gt "${TARGET_BLOCK_HEIGHT}" ]; then
      break;
    fi;
  fi;
	sleep 1;
done

docker exec -t --user postgres --workdir /var/lib/postgresql bcc-graphql_postgres_1 pg_dump -F c -d ${DB_NAME} -f ${FILE_NAME}
docker cp bcc-graphql_postgres_1:/var/lib/postgresql/${FILE_NAME} ${OUT_PATH}
printf "\nA new PostgreSQL dump has been written to ${OUT_PATH}\n"

scripts/export_env.sh
# Source this with the CONTEXT and NETWORK env set, containing the full path and
# supported network name.
# e.g.
# CONTEXT=$PWD NETWORK=mainnet . ./scripts/export_env.sh

export BIN_DIR=${CONTEXT}/bin
export CONFIG_DIR=${CONTEXT}/config/network/${NETWORK}
export SECRETS_DIR=${CONTEXT}/placeholder-secrets
export STATE_DIR=${CONTEXT}/state/network/${NETWORK}

case "$NETWORK" in
        mainnet)
            API_PORT=3100
            HASURA_PORT=8090
            METADATA_SERVER_URI="https://tokens.bcc.org"
            PG_ADMIN_PORT=8442
            POSTGRES_PORT=5442
            ;;
        testnet)
            API_PORT=3101
            HASURA_PORT=8091
            METADATA_SERVER_URI="https://metadata.bcc-testnet.tbcodev.io"
            OGMIOS_PORT=1338
            PG_ADMIN_PORT=8443
            POSTGRES_PORT=5443
            ;;
        aurum -qa)
            API_PORT=3102
            HASURA_PORT=8092
            METADATA_SERVER_URI="https://metadata.bcc-testnet.tbcodev.io"
            OGMIOS_PORT=1339
            PG_ADMIN_PORT=8444
            POSTGRES_PORT=5444
            ;;
          aurum -purple)
            API_PORT=3103
            HASURA_PORT=8093
            METADATA_SERVER_URI="https://metadata.bcc-testnet.tbcodev.io"
            OGMIOS_PORT=1340
            PG_ADMIN_PORT=8445
            POSTGRES_PORT=5445
            ;;
esac

export ALLOW_INTROSPECTION=true
export API_PORT
export BCC_NODE_CONFIG_PATH=${CONFIG_DIR}/bcc-node/config.json
export COMPOSE_DOCKER_CLI_BUILD=1
export DOCKER_BUILDKIT=1
export HASURA_CLI_PATH=${BIN_DIR}/hasura
export HASURA_PORT
export HASURA_URI=http://localhost:${HASURA_PORT}
export METADATA_SERVER_URI
export NETWORK
export OGMIOS_PORT
export POSTGRES_DB_FILE=${SECRETS_DIR}/postgres_db
export POSTGRES_PASSWORD_FILE=${SECRETS_DIR}/postgres_password
export PG_ADMIN_PORT
export POSTGRES_PORT
export POSTGRES_USER_FILE=${SECRETS_DIR}/postgres_user
export POSTGRES_HOST=localhost
